### PR TITLE
WIP: UEFI support

### PR DESCRIPTION
**Pull request by cab404**
Some notes: 
* I did not optimize amount of libraries, and added all of them.
* I observed some macbooks failing to boot with vesamenu, but I am not sure how to fix that without making sacrifices.

### CODE REVIEW

**Comment by cab404** [comment]
Also I don't know how to speak bat, so I need batman to fix bootinst.bat.

**Comment by cab404** [comment]
@Tomas-M ?

**Comment by Tomas-M** [comment]
Hello, thank you for your pull request. I can see you are asking for some response from my seid.
I didn't try it yet because I do not have any real device to test.

**Comment by cab404** [comment]
> I do not have any real device to test.

I think it should work just fine with vbox uefi. Also, do you know any way to make vesamenu fallback to menu on failure to load?

**Comment by Tomas-M** [comment]
I don't know

**Comment by Tomas-M** [comment]
What configuration file does this load when booted by EFI? Eg. syslinux.cfg, where it should be located?

**Comment by cab404** [comment]
> What configuration file does this load when booted by EFI? Eg. syslinux.cfg, where it should be located?

IIRC the same one as MBR, in /boot.

**Comment by cab404** [comment]
Found out that bootinst.sh copies EFI folder to a wrong location. So I will do some fixing and testing.

**Comment by felixlohmeier** [comment]
I tried to add files from this pull request to an existing bootable usb drive. The existing directory structure is:
* linux
  * boot

According to the changes in bootinst.sh the files should be placed here, right?
* linux
  * boot
    * efi
       * boot

This doesn't work for me. If I put the files from this pull request in the root directory...
* efi
  * boot
* linux
  * boot

... then I can boot into an UEFI shell but it shows a warning "no config file found" and I don't know how to actually boot into the operating system.

Could you please give an advice on how to test this pull request with an existing bootable usb drive?

**Comment by Tomas-M** [comment]
I would guess you will need

- efi
  - boot

- linux
   - boot

- boot
   - syslinux.cfg (copy from /linux/boot/syslinux.cfg)

**Comment by felixlohmeier** [comment]
Thanks @Tomas-M, I have tried your guess and finally came up with the following folder structure:
* EFI
  * boot
    * files from [folder efi-files in this pull request](https://github.com/cab404/linux-live/tree/master/bootfiles/efi-files)
    * renamed syslinux.efi to bootx64.efi
    * syslinux.cfg (copy from /linux/boot/syslinux.cfg)
* linux
  * boot

Using this setup the warning "Configuration file not found" is gone, but the screen remains blank even after waiting several minutes (while the cpu fan indicates that something is running).

If I delete the file /EFI/boot/syslinux.cfg the systems boots into the EFI shell with the warning "Configuration file not found", so the config file seems to be in the right place now.

@cab404: Any advice?

**Comment by cab404** [comment]
> Thanks @Tomas-M, I have tried your guess and finally came up with the following folder structure:
> 
> * EFI
>   
>   * boot
>     
>     * files from [folder efi-files in this pull request](https://github.com/cab404/linux-live/tree/master/bootfiles/efi-files)
>     * renamed syslinux.efi to bootx64.efi
>     * syslinux.cfg (copy from /linux/boot/syslinux.cfg)
> * linux
>   
>   * boot
> 
> Using this setup the warning "Configuration file not found" is gone, but the screen remains blank even after waiting several minutes (while the cpu fan indicates that something is running).
> 
> If I delete the file /EFI/boot/syslinux.cfg the systems boots into the EFI shell with the warning "Configuration file not found", so the config file seems to be in the right place now.
> 
> @cab404: Any advice?

Try changing vesamenu.c32 to menu.c32 in syslinux.cfg.

**Comment by felixlohmeier** [comment]
Thanks @cab404! Here is the full recipe for adding UEFI boot support to existing linux-live-kit installs with a workaround for vesamenu blank screen fail:

1. create folders for UEFI boot in root directory of usb drive
  * `EFI`
  * `EFI/boot`
2. copy files from current [syslinux release](https://mirrors.edge.kernel.org/pub/linux/utils/boot/syslinux/) (tested with syslinux-6.03) to usb drive
  * from `efi64/syslinux.efi` to `/EFI/boot/bootx64.efi`
  * from `efi64/com32/elflink/ldlinux/ldlinux.e64` to `/EFI/boot/ldlinux.e64`
  * from `efi64/com32/menu/menu.c32` to `/EFI/boot/menu.c32`
  * from `efi64/com32/libutil` to `/EFI/boot/libutil.c32`
3. create text file `/EFI/boot/syslinux.cfg`. Example:
```
UI menu.c32

TIMEOUT 40

LABEL default
MENU LABEL Run Linux (Persistent changes)
KERNEL /linux/boot/vmlinuz
APPEND vga=normal initrd=/linux/boot/initrfs.img load_ramdisk=1 prompt_ramdisk=0 nohd rw printk.time=0 consoleblank=0 slax.flags=perch apparmor=0

LABEL default
MENU LABEL Run Linux (Fresh start)
KERNEL /linux/boot/vmlinuz
APPEND vga=normal initrd=/linux/boot/initrfs.img load_ramdisk=1 prompt_ramdisk=0 nohd rw printk.time=0 consoleblank=0 apparmor=0

LABEL default
MENU LABEL Run Linux (Copy to RAM)
KERNEL /linux/boot/vmlinuz
APPEND vga=normal initrd=/linux/boot/initrfs.img load_ramdisk=1 prompt_ramdisk=0 nohd rw printk.time=0 consoleblank=0 slax.flags=toram apparmor=0
```

File: bootfiles/bootinst.sh [hunk 1]
```diff
@@ -40,5 +40,13 @@ if [ "$DEV" != "$PART" ]; then
...
 fi
 
+# Now to be able to boot from most of UEFI devices
+EFIDIR="$BOOT/EFI/boot"
+mkdir -p "$EFIDIR"
+cp efi-files/* "$EFIDIR"
+
+# I didn't rename the binary in place so it could be updated more easily
+mv "$EFIDIR/syslinux.efi" "$EFIDIR/bootx64.efi"
+
```
Comment (felixlohmeier) [inline bootfiles/bootinst.sh]: I would like to suggest to copy the syslinux.cfg file to the EFI directory as a starting point:
```
cp syslinux.cfg "$EFIDIR"
```

File: bootfiles/bootinst.sh [hunk 2]
```diff
@@ -40,5 +40,13 @@ if [ "$DEV" != "$PART" ]; then
    ) | fdisk $DEV >/dev/null 2>&1
 fi
 
+# Now to be able to boot from most of UEFI devices
+EFIDIR="$BOOT/EFI/boot"
```
Comment (felixlohmeier) [inline bootfiles/bootinst.sh]: EFI files need to be stored in the root directory of the usb drive. This line would result in EFIDIR=/linux/boot/EFI/boot.

I propose to change line 43 and 44 to (or something similar):
```
mkdir ../../EFI ../../EFI/boot
EFIDIR="$(readlink -f ../../EFI/boot)"
```

Comment (cab404) [reply bootfiles/bootinst.sh]: I think I did this because at some point `BOOT` pointed to usb root directory in Slax bootinst.

**Comment by mintsoft** [comment]
@felixlohmeier @cab404 @Tomas-M I've merged this and the description in https://github.com/Tomas-M/linux-live/pull/75#issuecomment-535481117 into https://github.com/mintsoft/linux-live/tree/uefi

I've moved all the file "logic" into the `./build` so that the Slax and the main `bootinst.sh`/`.bat` can be simple. 

I've just built this end-to-end and tested on a ProxMox VM + Dell XPS 13 in UEFI booting and it works a charm albeit with the non-vesa menu. 

I'm pretty sure the `vesamenu.c32` is broken in syslinux 6.03 as it always seems to throw this error even when provided with the correct build of `libcom32.c32`
```
undef symbol FAIL: __syslinux_debug_enabled
Failed to load libcom32.c32
Failed to load COM32 file vesamenu.c32
boot:
```

Any suggestions on that are welcomed!

**Comment by cab404** [comment]
I am closing this PR, merge #84 instead)

**Pull request closed without merging**